packages/playground/src/playground-db.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+defined('ABSPATH') || exit;
+
+function playground_escape_array($array)
+{
+	global $wpdb;
+	$escaped = array();
+	foreach ($array as $value) {
+		if (is_numeric($value))
+			$escaped[] = $wpdb->prepare('%d', $value);
+		else
+			$escaped[] = $wpdb->prepare('%s', $value);
+	}
+	return implode(',', $escaped);
+}
+
+function playground_dump_db($zip)
+{
+	global $wpdb;
+
+	$tables   = playground_get_db_tables();
+	$sql_dump = array();
+
+	foreach ($tables as $table) {
+		array_push(
+			$sql_dump,
+			sprintf("DROP TABLE IF EXISTS `%s`;", esc_sql($table)),
+			playground_dump_db_schema($table)
+		);
+	}
+
+	foreach ($tables as $table) {
+		$records = $wpdb->get_results(
+			sprintf(
+				'SELECT * FROM `%s`',
+				esc_sql($table)
+			),
+			ARRAY_A
+		);
+
+		if ($wpdb->last_error) {
+			error_log($wpdb->last_error);
+			continue;
+		}
+
+		foreach ($records as $record) {
+			array_push(
+				$sql_dump,
+				sprintf(
+					'INSERT INTO `%1$s` (%2$s) VALUES (%3$s);',
+					esc_sql($table),
+					playground_escape_array(
+						array_keys($record)
+					),
+					playground_escape_array(array_values($record))
+				)
+			);
+		}
+	}
+	$zip->addFile('schema/_Schema.sql', implode("\n", $sql_dump));
+
+}
+
+function playground_get_db_tables()
+{
+	global $wpdb;
+	$tables = $wpdb->get_results('SHOW TABLES', ARRAY_N);
+	return array_map(fn ($t) => $t[0], $tables);
+}
+
+function playground_dump_db_schema($table)
+{
+	global $wpdb;
+	$schema = $wpdb->get_row(
+		sprintf('SHOW CREATE TABLE `%s`', esc_sql($table)),
+		ARRAY_A
+	);
+	if ($wpdb->last_error) {
+		error_log($wpdb->last_error);
+		return '';
+	}
+	if (!isset($schema['Create Table'])) {
+		return '';
+	}
+	return preg_replace("/\s+/", " ", $schema['Create Table']);
```

Comment:
Won't this introduce unintended changes in statements like `CREATE TABLE test ( name varchar(255) default "FIRST      LAST" );`?

Reply:
The query breaks the dump if we don't remove the extra whitespace.
I tried limiting the replace by doing something like this `preg_replace("/([,(`])\s\s+/", "$1 ", $schema['Create Table'])` but the dump didn't work. I probably missed something in my regex, but couldn't figure it out.

If this isn't a big deal, I would prefer to skip it.

Reply:
Aha, I see! That's because we assume a single query per line. But that means we only have to remove newlines, not all whitespace, right?

Reply:
I just ran a very naive check of whether removing newlines could mess up the output of CREATE TABLE, and it seems like we're good in basic cases:

```
create table newline_test ( id int, name varchar(255) default "a
                   b" );
insert into newline_test (id) VALUES (1);
show create table newline_test;
```

```
CREATE TABLE `newline_test` ( `id` int(11) DEFAULT NULL, `name` varchar(255) DEFAULT 'a\n b' ) ENGINE=InnoDB DEFAULT CHARSET=latin1
```

In a future release, we could use the MySQL tokenizer shipped with the `sqlite-database-integration` plugin to normalize these queries not as strings but as semantic data structures. Either here, on write, or in Playground, on read.

Reply:
For more context – the "single query per line" constraint is the limitation of the SQL query parser shipped with that plugin. It knows how to read a single query, but it has no API to communicate "I reached the finishing semicolon and now I'm starting to read another query".

Reply:
Sometimes it's simple 🙂 
That regex worked, I updated it in ee10c8ff57078bc6c55942257163a8fa27c821ee and added a comment.